### PR TITLE
Add AppImage packaging to CI configuration

### LINKED ISSUE #944: Linux Appimage
Packaging as an [Appimage](https://appimage.org/) under Linux would be useful.

### PR DESCRIPTION
**Pull request by Technius**
Closes #944. Any reviews or suggestions are welcome since this is my first time packaging AppImages.

This will require some changes to the nightly release pipeline configuration (which renames package names containing `xournalpp-${VER}-` to some slightly different format; the generated AppImages are named `xournalpp-${VER}.AppImage`). It might be better to name the generated package `xournalpp-${VER}-${ARCH}.AppImage` to fix this configuration issue.

### CODE REVIEW

**Comment by Technius** [comment]
I updated the generated file name.

I cannot get the built AppImage file to run on my system (you can download it [here](https://dev.azure.com/technius/xournalpp/_build/results?buildId=40) by selecting the "Artifacts" dropdown menu). There seems to be some binary incompatibility issue with cairo:

```plain
(xournalpp-1.0.13.AppImage:4360): Gtk-WARNING **: 19:40:03.486: Could not load image
'pixmaps/com.github.xournalpp.xournalpp.svg': Unable to load image-loading module: /usr/lib64/gdk-
pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so: /usr/lib/librsvg-2.so.2: undefined symbol: 
cairo_tag_end
```

Cairo is packaged during the AppImage build (log [here](https://dev.azure.com/technius/xournalpp/_build/results?buildId=40), line 351), so I'm not sure why this error is popping up.

**Comment by LittleHuba** [comment]
I can run the AppImage on my Machine (Arch).
It seems you have not packaged a lot of icons though, which is what the error actually says. You need to include the Adwaita-Theme icons.

image

**Comment by Technius** [comment]
I updated the PR to bundle RSVG and some GDK Pixbuf files; this fixed the binary compatibility issues.

I don't think bundling an icon theme is the right approach. There should be some way to use a system icon theme.

**Comment by LittleHuba** [comment]
> I don't think bundling an icon theme is the right approach. There should be some way to use a system icon theme.

We are doing this for MacOS and Windows currently. If you find a better way we could review those releases as well.

**Comment by LittleHuba** [comment]
> I don't think bundling an icon theme is the right approach. There should be some way to use a system icon theme.

The thing is that AppImages are designed to be independent from the system. So relying on the system to have a compatible gtk-icon-theme installed kind of defeats the purpose.
What I have read so far is that you can include the icon theme but make GTK still search for locally installed themes and prefere those. But there were no specifics on how to do that.

File: azure-pipelines/util/appimage_apprun.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+#
+# Custom AppRun for Xournal++
+
+export PATH="$APPDIR/usr/bin/:$PATH"
```
Comment (TheAssassin) [inline azure-pipelines/util/appimage_apprun.sh]: Escaping here is suboptimal. Use `export PATH="$APPDIR"/usr/bin:"$PATH"`.

For the other uses, prefer quoting variables only as far as possible, non-special characters don't need quoting.

You should make sure `$APPDIR` is available when you e.g., want to use your script from an extracted AppImage. It won't run right now as-is.
Example code: `export APPDIR=${APPDIR:-$(readlink -f $(dirname "$0"))}`

File: azure-pipelines/util/appimage_apprun.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+export GDK_PIXBUF_MODULEDIR="${APPDIR}/usr/lib/gdk-pixbuf-2.0/loaders"
+export GDK_PIXBUF_MODULE_FILE="${APPDIR}/usr/lib/gdk-pixbuf-2.0/loaders.cache"
+export GTK_PATH="${APPDIR}/usr/lib/gtk-3.0"
+export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$GDK_PIXBUF_MODULEDIR"
```
Comment (TheAssassin) [inline azure-pipelines/util/appimage_apprun.sh]:
```suggestion
export LD_LIBRARY_PATH="$LD_LIBRARY_PATH":"$GDK_PIXBUF_MODULEDIR"
```

File: azure-pipelines/util/appimage_apprun.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+export PANGO_LIBDIR="$APPDIR/usr/lib"
```
Comment (TheAssassin) [inline azure-pipelines/util/appimage_apprun.sh]:
```suggestion
export PANGO_LIBDIR="$APPDIR"/usr/lib
```

File: azure-pipelines/util/appimage_apprun.sh [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+#
+# Custom AppRun for Xournal++
+
+export PATH="$APPDIR/usr/bin/:$PATH"
+export GDK_PIXBUF_MODULEDIR="${APPDIR}/usr/lib/gdk-pixbuf-2.0/loaders"
+export GDK_PIXBUF_MODULE_FILE="${APPDIR}/usr/lib/gdk-pixbuf-2.0/loaders.cache"
+export GTK_PATH="${APPDIR}/usr/lib/gtk-3.0"
```
Comment (TheAssassin) [inline azure-pipelines/util/appimage_apprun.sh]:
```suggestion
export GTK_PATH="$APPDIR"/usr/lib/gtk-3.0
```

File: azure-pipelines/util/appimage_apprun.sh [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+#
+# Custom AppRun for Xournal++
+
+export PATH="$APPDIR/usr/bin/:$PATH"
+export GDK_PIXBUF_MODULEDIR="${APPDIR}/usr/lib/gdk-pixbuf-2.0/loaders"
+export GDK_PIXBUF_MODULE_FILE="${APPDIR}/usr/lib/gdk-pixbuf-2.0/loaders.cache"
```
Comment (TheAssassin) [inline azure-pipelines/util/appimage_apprun.sh]:
```suggestion
export GDK_PIXBUF_MODULE_FILE="$APPDIR"/usr/lib/gdk-pixbuf-2.0/loaders.cache
```

File: azure-pipelines/util/appimage_apprun.sh [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+#
+# Custom AppRun for Xournal++
+
+export PATH="$APPDIR/usr/bin/:$PATH"
+export GDK_PIXBUF_MODULEDIR="${APPDIR}/usr/lib/gdk-pixbuf-2.0/loaders"
```
Comment (TheAssassin) [inline azure-pipelines/util/appimage_apprun.sh]:
```suggestion
export GDK_PIXBUF_MODULEDIR="$APPDIR"/usr/lib/gdk-pixbuf-2.0/loaders
```

File: azure-pipelines/util/appimage_apprun.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+export GDK_PIXBUF_MODULE_FILE="${APPDIR}/usr/lib/gdk-pixbuf-2.0/loaders.cache"
+export GTK_PATH="${APPDIR}/usr/lib/gtk-3.0"
+export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$GDK_PIXBUF_MODULEDIR"
+export PANGO_LIBDIR="$APPDIR/usr/lib"
+
+$APPDIR/usr/bin/xournalpp
```
Comment (TheAssassin) [inline azure-pipelines/util/appimage_apprun.sh]: Lacks quoting.

```suggestion
"$APPDIR"/usr/bin/xournalpp
```

File: azure-pipelines/util/build_appimage.sh [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+if [ x"${RSVG_LIBDIR}" != "x" ]; then
+    export GDK_PIXBUF_MODULE_FILE="${APPDIR}/usr/lib/gdk-pixbuf-2.0/loaders.cache"
+    echo "cp -a ${RSVG_LIBDIR}/librsvg*.so* $APPDIR/usr/lib"
+    cp -a "${RSVG_LIBDIR}"/librsvg*.so* "$APPDIR/usr/lib"
+fi
+
+set +x
+
+# Generate AppImage
+APPRUN="$(dirname $0)/appimage_apprun.sh"
```
Comment (TheAssassin) [inline azure-pipelines/util/build_appimage.sh]: `$(...)` should not be quoted.

File: readme/LinuxBuild.md [hunk 9]
```diff
@@ -105,7 +106,18 @@ TODO
 
 #### AppImage
 
-TODO
+One way to generate an AppImage is to first generate the `.tar.gz` package and
+then use that as the `AppDir` for `linuxdeploy`. For an example, see
+[`azure-pipelines/release.yml`](../azure-pipelines/release.yml)
+
+```bash
+cmake .. -DPACK_GENERATOR="TGZ"
```
Comment (TheAssassin) [inline readme/LinuxBuild.md]: Why not just use `make install DESTDIR=my.AppDir`? This seems a little too complex. Perhaps I'm missing something.

**Comment by TheAssassin** [comment]
Regarding your Gtk bundling, we started working on a plugin for linuxdeploy (see https://github.com/linuxdeploy/linuxdeploy-plugin-gtk). Perhaps you want to contribute to it to replace your own bundling code? That'll help others who want Gtk bundling, and will save you some code.

**Comment by Technius** [comment]
Thanks for the review!

> Regarding your Gtk bundling, we started working on a plugin for linuxdeploy (see https://github.com/linuxdeploy/linuxdeploy-plugin-gtk). Perhaps you want to contribute to it to replace your own bundling code? That'll help others who want Gtk bundling, and will save you some code.

I will look into it when I have time.

File: readme/LinuxBuild.md (hunk 9 above)
Comment (Technius) [reply readme/LinuxBuild.md]: The install target will actually install to `$DESTDIR/${CMAKE_INSTALL_PREFIX}`, which can cause problems if the user has configured cmake with `-DCMAKE_INSTALL_PREFIX=...`. The tarball seems to "just work" and I haven't found a better solution yet.

I also need to update the instructions here to use the `build_appimage.sh` script...

**Comment by Symbian9** [comment]
Any news? Really would be happy to use AppImage (as there no builds for Debian 9)

**Comment by Technius** [comment]
Unfortunately, I haven't had much time to work on this, but we're trying to get the Debian builds fixed by the end of this week. In the worst case, you can compile your own `.deb` package by following the [build instructions](https://github.com/xournalpp/xournalpp/blob/master/readme/LinuxBuild.md).

**Comment by TheAssassin** [comment]
@Symbian9 appimagecraft also works just fine, except for the translations not being updated properly (see #1617, workaround PR is in progress).

**Comment by Technius** [comment]
Finally had some more time to work on this, so I rebased and applied suggestions. I think it mostly works; I just need to test it on a docker image to check if I'm missing anything.

**Comment by Technius** [comment]
There were some lingering issues that I fixed, and it should be ready to go after some more testing.
* The linking issues had to be addressed by setting `LD_LIBRARY_PATH`.
* Newer systems have newer GTK themes that are generally incompatible with the version on GTK in the AppImage, so I decided to bundle Adwaita inside. Hopefully, this won't increase the file size too much.
* Out of caution, I also bundled the Adwaita icon theme. The system icon theme is used if possible, though (since I can't find a way to override the icon theme to Adwaita anyways).
* GTK uses `XDG_DATA_DIRS` to locate themes and icons so I had to set that in the `AppRun`.
* The `AppRun` now forwards command line arguments to `xournalpp`.

**Comment by Technius** [comment]
@Symbian9 Would you be willing to try the test AppImage? You can download it [here](https://dev.azure.com/technius/xournalpp/_build/results?buildId=247&view=artifacts&type=publishedArtifacts) under `linux-ubuntu-packages`. I have tried it on my system and it works, but I'd like to know if it works properly on other systems.

**Comment by TheAssassin** [comment]
That crappy page doesn't work on my system... mind to share a direct link?

**Comment by TheAssassin** [comment]
P.S.: https://github.com/TheAssassin/appimagelint

**Comment by Technius** [comment]
> That crappy page doesn't work on my system... mind to share a direct link?

Couldn't Inspect Element the download button directly (it's some JavaScript abomination) but [I managed to get a direct link](https://artprodcus3.artifacts.visualstudio.com/Aa25d95a8-7742-4e55-942f-fe2977ca150d/d07f160b-25db-4823-83e6-50b84810bac2/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3RlY2huaXVzL3Byb2plY3RJZC9kMDdmMTYwYi0yNWRiLTQ4MjMtODNlNi01MGI4NDgxMGJhYzIvYnVpbGRJZC8yNDcvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS1wYWNrYWdlcw2/content?format=file&subPath=%2Fxournalpp-1.1.0%2Bdev-x86_64.AppImage) in the request inspector.

> P.S.: https://github.com/TheAssassin/appimagelint

This looks very useful, I'll be trying it out shortly.

**Comment by TheAssassin** [comment]
Thanks. appimagelint is unhappy that the AppImage doesn't run on xenial (libstdc++ wise) and that some desktop files cannot be validated using desktop-file-utils. The AppImage starts just fine, though. That's great!

The theme is a bit off (size of elements, buttons have no spacing to the dialog edges, etc.), perhaps we can fix that at some point.

Any reason to abandon appimagecraft by the way?

**Comment by Technius** [comment]
> appimagelint is unhappy that the AppImage doesn't run on xenial (libstdc++ wise)

We moved to C++17 for master branch so this is unavoidable; we'll probably drop xenial in favor of bionic. The `1.0.x` branch is still on C++11 and can probably still be built on xenial without any newer compilers (need to confirm this though).

> The theme is a bit off (size of elements, buttons have no spacing to the dialog edges, etc.), perhaps we can fix that at some point.

Assuming you're not already using Adwaita, maybe that's just how Adwaita looks? I tried using my system GTK theme, but it is not backwards compatible with GTK 3.18 and ends up looking horrible.

> Any reason to abandon appimagecraft by the way?

Not really abandoned--just haven't tried it yet.

---

I realized that we might be missing some of the GTK input method module binaries. Whether that is an issue will become apparent in the future.

**Comment by TheAssassin** [comment]
I'm on Xfce, not a fan of GNOME 3, so I expected some differences.

It's IMO fine to have support for bionic and newer only, if there's a significant interest in support for older distros you can still implement static linking, I guess.

Regarding appimagelint's other errors, please check if you can fix them.

By the way, some of your fixes (e.g., the environment variables, etc.) could be added to https://github.com/linuxdeploy/linuxdeploy-plugin-gtk. Contributions welcome!

**Comment by alfredoivan** [comment]
> > That crappy page doesn't work on my system... mind to share a direct link?
> 
> Couldn't Inspect Element the download button directly (it's some JavaScript abomination) but [I managed to get a direct link](https://artprodcus3.artifacts.visualstudio.com/Aa25d95a8-7742-4e55-942f-fe2977ca150d/d07f160b-25db-4823-83e6-50b84810bac2/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3RlY2huaXVzL3Byb2plY3RJZC9kMDdmMTYwYi0yNWRiLTQ4MjMtODNlNi01MGI4NDgxMGJhYzIvYnVpbGRJZC8yNDcvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS1wYWNrYWdlcw2/content?format=file&subPath=%2Fxournalpp-1.1.0%2Bdev-x86_64.AppImage) in the request inspector.
> 
> > P.S.: https://github.com/TheAssassin/appimagelint
> 
> This looks very useful, I'll be trying it out shortly.

Hello. I tried using the appimage in ubuntu trusty 14.04.6
It failed. Here is the log: https://pastebin.com/PK8ZgWL3

Although no expert in appimage, it seems like these two libraries would suffice:
CXXABI_1.3.9
GLIBCXX_3.4.22

**Comment by TheAssassin** [comment]
@alfredoivan please read my [last comment](https://github.com/xournalpp/xournalpp/pull/1448#issuecomment-612376082) or [@Technius's comment](https://github.com/xournalpp/xournalpp/pull/1448#issuecomment-612330266):

> It's IMO fine to have support for bionic and newer only, if there's a significant interest in support for older distros you can still implement static linking, I guess.

Ubuntu 18.04 and newer are supported only.
Ubuntu 14.04 is end-of-life anyway, it does not receive any updates any more. Please upgrade to Ubuntu 18.04.

**Comment by Technius** [comment]
> Regarding appimagelint's other errors, please check if you can fix them.

Those have to do with the mime type files (#1866) and will be dealt with later.

**Comment by Technius** [comment]
Updated the readme. WIll merge this tomorrow unless there's something missing.

**Pull request merged**